### PR TITLE
[MXNET-411] Add ROI Align

### PR DESCRIPTION
**Pull request by zhanghang1989**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] ROI Align from Caffe2
- [x] docs and unit test

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by piiswrong** [comment]
Please add tests and documentation

**Comment by xinyu-intel** [comment]
Hi, can you enable openmp on the cpu implementation of roialign. This can achieve a better performance. You can reference my pr in #9958.

**Comment by wkcn** [comment]
The ROIAlign in this PR is ROIAlign(max).
And the ROIAlign in Caffe2([forward](https://github.com/caffe2/caffe2/blob/master/caffe2/operators/roi_align_op.cu), [backward](https://github.com/caffe2/caffe2/blob/master/caffe2/operators/roi_align_gradient_op.cu)) is ROIAlign(mean)

**Comment by wkcn** [comment]
Hi. I wrote a ROIAlign forward/backward test. It may be useful.
https://github.com/wkcn/MobulaOP/blob/master/tests/test_roi_align_op.py

And I found the implenent of CPU/GPU is different. I think it's better to add cpu/gpu consistence test.

**Comment by zhanghang1989** [comment]
The docs are generated here http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-10852/9/api/python/ndarray/contrib.html?highlight=roi#mxnet.ndarray.contrib.ROIAlign

**Comment by zhanghang1989** [comment]
@piiswrong @zhreshold Could you please review this PR for adapting ROI Aligh from Caffe2. Thanks!

File: tests/python/unittest/test_operator.py [hunk 1]
```diff
@@ -6018,6 +6019,24 @@ def test_context_num_gpus():
...
+    x1 = np.random.rand(4, 1, 12, 12).astype('float64')
+    x2 = np.array([[0, 1.1, 1.1, 6.2, 6.2], [2, 6.1, 2.1, 8.2, 11.2], [1, 3.1, 1.1, 5.2, 10.2]], dtype='float64')
+
+    check_numeric_gradient(sym=test, location=[x1, x2],
+                           grad_nodes={'data':'write', 'rois':'null'},
+                           numeric_eps=1e-4, rtol=1e-1, atol=1e-4)
+    check_numeric_gradient(sym=test, location=[x1, x2],
+                           grad_nodes={'data':'add', 'rois':'null'},
+                           numeric_eps=1e-4, rtol=1e-1, atol=1E-4)
+
```
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply tests/python/unittest/test_operator.py]: Yeah, will a forward result check soon 👍

**Comment by zhanghang1989** [comment]
@piiswrong @zhreshold I have added the unit tests. Please see the updates. Thanks!

File: src/operator/contrib/adaptive_avg_pooling.cc [hunk 2]
```diff
@@ -26,8 +26,8 @@
 // #include "elemwise_op_common.h"
 #include "../elemwise_op_common.h"
 
-#define START_IND(a, b, c) static_cast<int>(floor(static_cast<float>(a * c) / b))
-#define END_IND(a, b, c) static_cast<int>(ceil(static_cast<float>((a + 1) * c) / b))
+#define START_IND(a, b, c) static_cast<int>(floor(static_cast<real>(a * c) / b))
```
Comment (piiswrong) [inline src/operator/contrib/adaptive_avg_pooling.cc]: ?

File: src/operator/contrib/roi_align.cc [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+      roi_batch_ind = offset_bottom_rois[0];
+      offset_bottom_rois++;
+    }
+
+    // Do not using rounding; this implementation detail is critical
+    T roi_start_w = offset_bottom_rois[0] * spatial_scale;
+    T roi_start_h = offset_bottom_rois[1] * spatial_scale;
+    T roi_end_w = offset_bottom_rois[2] * spatial_scale;
+    T roi_end_h = offset_bottom_rois[3] * spatial_scale;
+    // T roi_start_w = round(offset_bottom_rois[0] * spatial_scale);
```
Comment (piiswrong) [inline src/operator/contrib/roi_align.cc]: ?

File: src/operator/contrib/roi_align.cc [hunk 4]
```diff
@@ -0,0 +1,593 @@
...
+    const int width,
+    const int pooled_height,
+    const int pooled_width,
+    const int sampling_ratio,
+    T* bottom_diff,
+    const T* bottom_rois,
+    int rois_cols) {
+  DCHECK(rois_cols == 4 || rois_cols == 5);
+
+  for (int index = 0; index < nthreads; index++) {
```
Comment (piiswrong) [inline src/operator/contrib/roi_align.cc]: use openmp?

File: src/operator/contrib/roi_align.cc [hunk 5]
```diff
@@ -0,0 +1,593 @@
...
+    const T* bottom_rois,
+    int roi_cols,
+    T* top_data) {
+  DCHECK(roi_cols == 4 || roi_cols == 5);
+
+  int n_rois = nthreads / channels / pooled_width / pooled_height;
+  // (n, c, ph, pw) is an element in the pooled output
+  // can be parallelized using omp
+  // #pragma omp parallel for num_threads(32)
+  for (int n = 0; n < n_rois; n++) {
```
Comment (piiswrong) [inline src/operator/contrib/roi_align.cc]: openmp?

File: src/operator/contrib/roi_align.cu [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+  Stream<gpu> *s = ctx.get_stream<gpu>();
+  cudaStream_t stream = mshadow::Stream<gpu>::GetStream(s);
+
+  // assume all the data and gradient have the same type
+  MSHADOW_REAL_TYPE_SWITCH(out_grad[0].type_flag_, DType, {
+    const DType *top_diff = out_grad[0].dptr<DType>();
+    const DType *bottom_rois = in_data[0].dptr<DType>();
+    DType *grad_in = outputs[0].dptr<DType>();
+
+    if (kAddTo == req[roialign::kData] || kWriteTo == req[roialign::kData]) {
```
Comment (piiswrong) [inline src/operator/contrib/roi_align.cu]: Return if NullOp before the switch?

File: src/operator/contrib/roi_align-inl.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+}  // roialign
+
+
+struct ROIAlignParam : public dmlc::Parameter<ROIAlignParam> {
+  TShape pooled_size;
+  float spatial_scale;
+  DMLC_DECLARE_PARAMETER(ROIAlignParam) {
+    DMLC_DECLARE_FIELD(pooled_size)
+    .set_expect_ndim(2).enforce_nonzero()
+    .describe("fix pooled size: (h, w)");
```
Comment (piiswrong) [inline src/operator/contrib/roi_align-inl.h]: what's pooled_size?

File: src/operator/contrib/roi_align-inl.h [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    DMLC_DECLARE_FIELD(spatial_scale).set_range(0.0, 1.0)
+    .describe("Ratio of input feature map height (or w) to raw image height (or w). "
+    "Equals the reciprocal of total stride in convolutional layers");
+  }
+};
+
+
+struct ROIAlignGrad {
```
Comment (piiswrong) [inline src/operator/contrib/roi_align-inl.h]: No need for this struct. Use lambda directly at set_attr

File: src/operator/contrib/roi_align-inl.h (hunk 7 above)
Comment (zhanghang1989) [reply src/operator/contrib/roi_align-inl.h]: The output roi feature sizes. Name is compatible with ROIPooling

File: src/operator/contrib/roi_align.cc [hunk 9]
```diff
@@ -0,0 +1,586 @@
...
+    const T* bottom_rois,
+    int roi_cols,
+    T* top_data) {
+  DCHECK(roi_cols == 4 || roi_cols == 5);
+
+  int n_rois = nthreads / channels / pooled_width / pooled_height;
+  // (n, c, ph, pw) is an element in the pooled output
+  // can be parallelized using omp
+  int n;
+#pragma omp parallel for private(n) \
```
Comment (xinyu-intel) [inline src/operator/contrib/roi_align.cc]: Thanks for adding omp. Regarding to `NUM_OF_ROIS` and `CHANNELS`, I found laster is more and more bigger than former in ROIPooling, so I just apply omp on channels to achieve better application performance. Can you help benchmark the performance based on the usually roi_align size?

File: src/operator/contrib/roi_align.cc [hunk 10]
```diff
@@ -0,0 +1,586 @@
...
+        roi_start_h,
+        roi_start_w,
+        bin_size_h,
+        bin_size_w,
+        roi_bin_grid_h,
+        roi_bin_grid_w,
+        &pre_calc);
+
+    int c;
+#pragma omp parallel for private(c) \
```
Comment (xinyu-intel) [inline src/operator/contrib/roi_align.cc]: same

File: src/operator/contrib/roi_align.cc (hunk 9 above)
Comment (zhanghang1989) [reply src/operator/contrib/roi_align.cc]: Are you suggesting removing this omp?

File: src/operator/contrib/roi_align.cc (hunk 10 above)
Comment (zhanghang1989) [reply src/operator/contrib/roi_align.cc]: keeping this ?

**Comment by zhanghang1989** [comment]
Remove OMP in backward, due to no atomic add in cpu.

**Comment by zhreshold** [comment]
Gathering status, did anyone have unresolved issues?
Trying to merge this PR ASAP as it is the required operator for gluon-cv.

File: src/operator/contrib/roi_align-inl.h [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+}  // roialign
+
+
+struct ROIAlignParam : public dmlc::Parameter<ROIAlignParam> {
+  TShape pooled_size;
+  float spatial_scale;
+  DMLC_DECLARE_PARAMETER(ROIAlignParam) {
+    DMLC_DECLARE_FIELD(pooled_size)
+    .set_expect_ndim(2).enforce_nonzero()
+    .describe("ROI Align output roi featuremap height and width: (h, w)");
```
Comment (piiswrong) [inline src/operator/contrib/roi_align-inl.h]: featuremap -> feature map

File: src/operator/contrib/roi_align.cc [hunk 12]
```diff
@@ -0,0 +1,580 @@
...
+      Fill<false>(s, outputs[1], kWriteTo, static_cast<DType>(0));
+    }
+  })
+}
+
+DMLC_REGISTER_PARAMETER(ROIAlignParam);
+
+NNVM_REGISTER_OP(_contrib_ROIAlign)
+.describe(R"code(
+ROI Align Layer
```
Comment (piiswrong) [inline src/operator/contrib/roi_align.cc]: Superfluous. Remove this line

**Comment by zhanghang1989** [comment]
I should have addressed the comments. Please let me know if there are any further notes. Thanks

**Comment by zhanghang1989** [comment]
Finally it passes the CI :) @piiswrong

File: src/operator/contrib/roi_align.cu [hunk 13]
```diff
@@ -0,0 +1,484 @@
...
+         stream>>>(
+          count,
+          bottom_data,
+          param.spatial_scale,
+          channels,
+          height,
+          width,
+          pooled_height,
+          pooled_width,
+          -1,
```
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply src/operator/contrib/roi_align.cu]: Currently, it uses the adaptive size. I will make it as an option

**Pull request merged**